llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1565,10 +1567,17 @@ static void transformRecipestoEVLRecipes(VPlan &Plan, VPValue &EVL) {
         VPValue *CurVPV = CurRecipe->getVPSingleValue();
         CurVPV->replaceAllUsesWith(NewRecipe->getVPSingleValue());
       }
-      CurRecipe->eraseFromParent();
+      // Defer erasing recipes till the end so that we don't invalidate the
+      // VPTypeAnalysis cache
+      ToErase.push_back(CurRecipe);
     }
-    recursivelyDeleteDeadRecipes(HeaderMask);
```

Comment:
Can you add `HeaderMask` to be erased here?

Reply:
Is it possible that `HeaderMask` might still be used by a recipe if it can't be converted to an EVL recipe?

Reply:
In that case, maybe just recursively delete the recipes to erase, which then will automatically clean up the header mask if it becomes dead?

All recipes you are queuing should be `VPSingelDefRecipe*`

```
 for (VPRecipeBase *R : ToErase)
     recursivelyDeleteDeadRecipes(R);
```

Reply:
Some of the recipes are stores with zero values, but I was able to copy the trick from optimizeForVFAndUF where you erase the recipe and then call recursivelyDeleteDeadRecipes on its operands